common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,361 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.net;
+
+import com.microsoft.identity.common.java.telemetry.Telemetry;
+import com.microsoft.identity.common.java.telemetry.events.HttpEndEvent;
+import com.microsoft.identity.common.java.telemetry.events.HttpStartEvent;
+import com.microsoft.identity.common.java.util.StringUtil;
+import com.microsoft.identity.common.java.util.ported.Consumer;
+import com.microsoft.identity.common.java.util.ported.Function;
+import com.microsoft.identity.common.java.util.ported.Supplier;
+
+import net.jcip.annotations.Immutable;
+import net.jcip.annotations.ThreadSafe;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.SocketTimeoutException;
+import java.net.URL;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicReference;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.NonNull;
+
+import static com.microsoft.identity.common.java.AuthenticationConstants.AAD.CLIENT_REQUEST_ID;
+
+/**
+ * A client object for handling HTTP requests and responses.  This class accepts a RetryPolicy that
+ * is applied to the responses.  By default, the policy is a null policy not retrying anything.  The
+ * default settings for read timeout and connection timeout are 30s, and the default setting for the
+ * size of the buffer for reading objects from the http stream is 1024 bytes.
+ * <p>
+ * There are two ways to supply timeout values to this class, one method takes suppliers, the other
+ * one integers.  If you use both of these, the suppliers method will take precendence over the method
+ * using integers.
+ */
+@AllArgsConstructor
+@Builder
+@ThreadSafe
+@Immutable
+public class UrlConnectionHttpClient extends AbstractHttpClient {
```

Comment:
>UrlConnectionHttpClient [](http://example.com/codeflow?start=13&length=23)

I am not seeing any telemetry for exceptions/errors in this class... unless it is getting captured in someotherway..
It is very important that we capture telemetry for different kinds of failures here.. to make debugging with logs easier

Reply:
We can add that later. (We don't have a good telemetry story yet tbh).

I'll leave a todo here.